File: windows/KikuKegel.Windows/DailyCheckInStore.cs
using System.Text.Json;

namespace KikuKegel.Windows;

internal sealed class DailyCheckInSummary
{
    public required bool[] CompletedSlots { get; init; }
    public required int TodayCount { get; init; }
    public required int YesterdayCount { get; init; }
    public int TotalCount => CompletedSlots.Length;

    public string ComparisonText
    {
        get
        {
            var delta = (int)Math.Round((TodayCount - YesterdayCount) / (double)TotalCount * 100);
            if (delta > 0)
            {
                return $"比昨天 +{delta}%，真棒";
            }

            if (delta < 0)
            {
                return $"比昨天 {delta}%，再接再厉";
            }

            return "和昨天持平，继续保持";
        }
    }
}

internal sealed class DailyCheckInStore
{
    public const int SlotCount = 15;
    private readonly string _storagePath = StoragePaths.File("daily-check-in.json");
    private Dictionary<string, HashSet<int>> _records;

    public DailyCheckInStore()
    {
        _records = Load();
    }

    public bool MarkCompletion(DateTime? date = null, int? preferredSlotIndex = null)
    {
        var now = date ?? DateTime.Now;
        var key = DateKey(now);
        var slotIndex = preferredSlotIndex.HasValue
            ? Math.Clamp(preferredSlotIndex.Value, 0, SlotCount - 1)
            : NearestSlotIndex(now);

        if (!_records.TryGetValue(key, out var slots))
        {
            slots = new HashSet<int>();
            _records[key] = slots;
        }

        if (!slots.Add(slotIndex))
        {
            return false;
        }

        Save();
        return true;
    }

    public int NearestSlotIndex(DateTime? date = null)
    {
        var now = date ?? DateTime.Now;
        var firstSlot = now.Date.AddHours(9);
        var lastSlot = firstSlot.AddMinutes((SlotCount - 1) * 45);

        if (now <= firstSlot)
        {
            return 0;
        }

        if (now >= lastSlot)
        {
            return SlotCount - 1;
        }

        var minutes = Math.Abs((now - firstSlot).TotalMinutes);
        return Math.Clamp((int)Math.Round(minutes / 45), 0, SlotCount - 1);
    }

    public DailyCheckInSummary Summary(DateTime? date = null)
    {
        var now = date ?? DateTime.Now;
        var todayKey = DateKey(now);
        var yesterdayKey = DateKey(now.AddDays(-1));
        var todaySlots = _records.TryGetValue(todayKey, out var today) ? today : new HashSet<int>();
        var yesterdaySlots = _records.TryGetValue(yesterdayKey, out var yesterday) ? yesterday : new HashSet<int>();

        return new DailyCheckInSummary
        {
            CompletedSlots = Enumerable.Range(0, SlotCount).Select(todaySlots.Contains).ToArray(),
            TodayCount = todaySlots.Count,
            YesterdayCount = yesterdaySlots.Count
        };
    }

    private Dictionary<string, HashSet<int>> Load()
    {
        try
        {
            if (!File.Exists(_storagePath))
            {
                return new Dictionary<string, HashSet<int>>();
            }

            var decoded = JsonSerializer.Deserialize<Dictionary<string, int[]>>(File.ReadAllText(_storagePath));
            return decoded?.ToDictionary(
                pair => pair.Key,
                pair => pair.Value.Where(index => index >= 0 && index < SlotCount).ToHashSet()
            ) ?? new Dictionary<string, HashSet<int>>();
        }
        catch
        {
            return new Dictionary<string, HashSet<int>>();
        }
    }

    private void Save()
    {
        var encoded = _records.ToDictionary(pair => pair.Key, pair => pair.Value.OrderBy(index => index).ToArray());
        File.WriteAllText(_storagePath, JsonSerializer.Serialize(encoded));
    }

    private static string DateKey(DateTime date) => date.ToString("yyyy-MM-dd");
}

File: Sources/KikuKegel/DailyCheckInStore.swift
import Foundation

struct DailyCheckInSummary {
    let completedSlots: [Bool]
    let todayCount: Int
    let yesterdayCount: Int

    var totalCount: Int {
        completedSlots.count
    }

    var comparisonText: String {
        let delta = Int(round((Double(todayCount - yesterdayCount) / Double(totalCount)) * 100))
        if delta > 0 {
            return "比昨天 +\(delta)%，真棒"
        }
        if delta < 0 {
            return "比昨天 \(delta)%，再接再厉"
        }
        return "和昨天持平，继续保持"
    }
}

final class DailyCheckInStore {
    static func slotCount(
        for scheduleMode: ReminderScheduleMode,
        reminderInterval: KegelReminderInterval = .fortyFive
    ) -> Int {
        scheduleMode.slotMinutes(interval: reminderInterval.minutes).count
    }

    private let storageKey = "DailyCheckInRecords.v1"
    private let defaults: UserDefaults
    private var records: [String: Set<Int>]

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.records = Self.loadRecords(from: defaults, key: storageKey)
    }

    func markCompletion(
        at date: Date = Date(),
        preferredSlotIndex: Int? = nil,
        scheduleMode: ReminderScheduleMode = .weekdays,
        reminderInterval: KegelReminderInterval = .fortyFive
    ) -> Bool {
        let key = dateKey(for: date)
        let count = Self.slotCount(for: scheduleMode, reminderInterval: reminderInterval)
        let slotIndex = normalizedSlotIndex(preferredSlotIndex, count: count)
            ?? nearestSlotIndex(for: date, scheduleMode: scheduleMode, reminderInterval: reminderInterval)
        var slots = records[key, default: []]

        guard !slots.contains(slotIndex) else {
            return false
        }

        slots.insert(slotIndex)
        records[key] = slots
        save()
        return true
    }

    func nearestSlotIndex(
        for date: Date = Date(),
        scheduleMode: ReminderScheduleMode = .weekdays,
        reminderInterval: KegelReminderInterval = .fortyFive
    ) -> Int {
        let slots = slotDates(for: date, scheduleMode: scheduleMode, reminderInterval: reminderInterval)
        guard let firstSlot = slots.first, let lastSlot = slots.last else {
            return 0
        }

        if date <= firstSlot {
            return 0
        }
        if date >= lastSlot {
            return max(0, slots.count - 1)
        }

        var nearestIndex = 0
        var nearestDistance = abs(date.timeIntervalSince(firstSlot))
        for (index, slotDate) in slots.enumerated().dropFirst() {
            let distance = abs(date.timeIntervalSince(slotDate))
            if distance < nearestDistance {
                nearestIndex = index
                nearestDistance = distance
            }
        }
        return nearestIndex
    }

    func summary(
        for date: Date = Date(),
        scheduleMode: ReminderScheduleMode = .weekdays,
        reminderInterval: KegelReminderInterval = .fortyFive
    ) -> DailyCheckInSummary {
        let todayKey = dateKey(for: date)
        let yesterday = Calendar.current.date(byAdding: .day, value: -1, to: date) ?? date
        let yesterdayKey = dateKey(for: yesterday)
        let todaySlots = records[todayKey, default: []]
        let yesterdaySlots = records[yesterdayKey, default: []]
        let count = Self.slotCount(for: scheduleMode, reminderInterval: reminderInterval)
        let completedSlots = (0..<count).map { todaySlots.contains($0) }

        return DailyCheckInSummary(
            completedSlots: completedSlots,
            todayCount: todaySlots.filter { $0 < count }.count,
            yesterdayCount: yesterdaySlots.filter { $0 < count }.count
        )
    }

    private func slotDates(
        for date: Date,
        scheduleMode: ReminderScheduleMode,
        reminderInterval: KegelReminderInterval
    ) -> [Date] {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        return scheduleMode.slotMinutes(interval: reminderInterval.minutes).compactMap { minute in
            calendar.date(byAdding: .minute, value: minute, to: startOfDay)
        }
    }

    private func normalizedSlotIndex(_ index: Int?, count: Int) -> Int? {
        guard let index else { return nil }
        return min(max(index, 0), max(0, count - 1))
    }

    private func dateKey(for date: Date) -> String {
        let components = Calendar.current.dateComponents([.year, .month, .day], from: date)
        let year = components.year ?? 0
        let month = components.month ?? 0
        let day = components.day ?? 0
        return String(format: "%04d-%02d-%02d", year, month, day)
    }

    private func save() {
        let encodedRecords = records.mapValues { Array($0).sorted() }
        guard let data = try? JSONEncoder().encode(encodedRecords) else {
            return
        }
        defaults.set(data, forKey: storageKey)
    }

    private static func loadRecords(from defaults: UserDefaults, key: String) -> [String: Set<Int>] {
        guard
            let data = defaults.data(forKey: key),
            let decoded = try? JSONDecoder().decode([String: [Int]].self, from: data)
        else {
            return [:]
        }

        return decoded.mapValues { indexes in
            Set(indexes.filter { 0..<64 ~= $0 })
        }
    }
}
